src/coprocessor/dag/expr/builtin_compare.rs

```diff
@@ -202,6 +202,60 @@ impl ScalarFunc {
     pub fn in_json(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
         do_in(self, |v| v.eval_json(ctx, row), |l, r| Ok(l.cmp(r)))
     }
+
+    pub fn interval_int(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        let target = match self.children[0].eval_int(ctx, row) {
+            Err(e) => return Err(e),
+            Ok(None) => return Ok(Some(-1)),
+            Ok(Some(v)) => v,
+        };
+        let tus = mysql::has_unsigned_flag(self.children[0].get_tp().get_flag());
+
+        let mut left = 1;
+        let mut right = self.children.len();
+        while left < right {
+            let mid = left + (right - left) / 2;
+            let m = self.children[mid].eval_int(ctx, row)?.unwrap_or(target);
+
+            let mus = mysql::has_unsigned_flag(self.children[mid].get_tp().get_flag());
+
+            let cmp = match (tus, mus) {
+                (false, false) => target < m,
+                (true, true) => (target as u64) < (m as u64),
+                (false, true) => target < 0 || (target as u64) < (m as u64),
+                (true, false) => m > 0 && (target as u64) < (m as u64),
+            };
+
+            if !cmp {
+                left = mid + 1
+            } else {
+                right = mid
+            }
+        }
+        Ok(Some((left - 1) as i64))
+    }
+
+    pub fn interval_real(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        let target = match self.children[0].eval_real(ctx, row) {
+            Err(e) => return Err(e),
+            Ok(None) => return Ok(Some(-1)),
+            Ok(Some(v)) => v,
+        };
+
+        let mut left = 1;
```

Comment:
Why our initial left is 1 while in TiDB i starts from 0?

Reply:
The code at TiDB side makes a slice that starts from 1 when passing args to `binSearch`:

https://github.com/pingcap/tidb/blob/cef80b3da390b8b35c703546c54cda9a251a18de/expression/builtin_compare.go#L841

The code here initialize `left` as 1 and return `left - 1` as result, which should be the same as TiDB code. @breeswish